### ISSUE TITLE
Distorted sound when looping the first bar

### ISSUE THREAD

**Issue by michaelgregorius** [LMMS/lmms#2539]
Steps to reproduce:
* Start LMMS.
* Add an A2 note that's three bars long to the Triple Oscillator track.
* Activate looping in the Song Editor. Keep the default of only looping the first bar.
* Hit "Play" in the Song Editor.

The sound becomes distorted when the first loop occurs and even more distorted when looping for the second time. When you set the loop length to two bars the distortion disappears in the middle of the loop. When you set the loop length to three bars the problem disappears. This affects all instruments.

**Comment by grejppi** [comment]
If a note has been set to play for three bars, it *will* play for three bars. Looping won't cut them, so in your use case there will be multiple notes at the same time. And multiple notes, each by default at full volume (#113), will obviously clip and distort the sound.

Is this about the volume of the notes, or the fact that they will not be cut off?

**Comment by michaelgregorius** [comment]
It's about the fact that the note is not cut off. I have just tested how REAPER behaves in these cases and it cuts off the note by sending a "Note Off" event for the corresponding note at the end of the loop. I assume that most other DAWs will do the same.

I also noticed that LMMS does not play the sound of a track if its "Note On" event starts before the loop.

These behaviours make it very hard to work on a section of a song by looping and adjusting it iteratively for the following reasons:
* Missing note cut off: After the first loop you might hear sounds in a part that do not belong there, i.e. that do not occur there if you play the song from start to finish without looping. Example: a track plays a single note from the start of bar 3 to the end of bar 4. The loop is set from the start of bar 1 to the end of bar 3. After the first loop you will hear the tail of the instrument at the beginning of the song.
* Missing "Note On": You won't hear instruments that start before the loop and that would normally play through the looped section. Example: You have some pads from the start of bar 5 to the end of bar 8. You loop from the start of bar 7 to the end of bar 8. The pad sound will not be played.

So both problems give a false picture of what is happening in the section of a loop. To solve the first problem LMMS would need to implement something called "MIDI Chasing" where you check for MIDI events that have happened before the starting point of playback and that will affect the section that is currently played back.

**Comment by tresf** [comment]
Reminds me in some ways of #1287.

**Comment by Wallacoloo** [comment]
@michaelgregorius the missing "Note On" events have no graceful solution. There are plenty of effects which have infinite impulse response (IIR; i.e. an input can create a sound that never fully decays). Consider a delay effect with feedback=100%, for example, or even an instrument in which the release parameter is such that the note never actually becomes inaudible.

Even if you can determine which notes *should* be active, and synthesize their note-on events during a loop, you would have to walk all the way back to the start of the song & play *every* note in the song in order to accomodate for any IIR effects. Even if you disregarded IIR events (which is a somewhat valid approach), you would still have to render & discard the beginning portion of any notes which start before the playhead just due to their ADSR envelopes. Therefore, I think the current "note on" behavior is OK.

On the other hand, I think notes *should* be released at the end of a loop. In fact, their output should likely be truncated at the end of a loop (i.e. make looping equivalent to pressing "play", then "stop" [not "pause"], seeking to the start of the loop, and pressing "play" again. Semantically, this is what "looping" means to me).

The behavior I described (no midi chasing, but yes to releasing notes at the end of a loop) is the same behavior used by Bitwig Studio. I have no other DAWs to test on though.

**Comment by musikBear** [comment]
maby #2047 is a candidate relations

**Comment by Umcaruje** [comment]
>maby #2047 is a candidate relations

This has to do with notes not getting a noteoff when looping, I can't see any connection with that problem.

>The behavior I described (no midi chasing, but yes to releasing notes at the end of a loop) is the same behavior used by Bitwig Studio. 

Yeah, I use Bitwig, and must say that I never had a need for MIDI chasing when looping in it. 

So I agree with @Wallacoloo that notes should be cut off at the ends of the loops. The MIDI Chasing is a bit too complex and there are many places where it could go wrong.

**Comment by michaelgregorius** [comment]
@Wallacoloo I agree that it's definitively not feasible to account for everything that has happened before the position of the playhead. In such a scenario if you had the playhead near the end of the song this would mean that in the worst case you'd have to render almost the complete song just to get an accurate picture.

I have checked what REAPER does for the "Note On before Loop Start" cases and it seems that it just acts as if a "Note On" event takes place at the start of the loop. In my opinion such a solution is better than the current behaviour as it will give a better impression because there are no sounds missing (albeit the sounds played may be incorrect due to modulations that might be triggered at the wrong place). One solution might be to provide an option to the user whether such a track should stay silent or whether a "Note On" event should be triggered.

#2047 is indeed not related to this problem but is about precision problems with some (all?) of LMMS' oscillators due to using `float` types instead of `double` types for the phase calculation.

**Comment by Wallacoloo** [comment]
@michaelgregorius if we implement the "note on at loop start" behavior, I
propose that the same thing should happen during a normal, non-looping
play, such that looping remains semantically equivalent to repeated
stopping, seeking and playing.

Though I have no strong opposition to the idea of making this
user-configurable, my concern is in how the option would then be presented
from a UX point of view. I don't think it will receive enough use to
justify having a dedicated button for it on the UI. We could put it in the
settings menu, but this doesn't seem like a user preference, but rather
something that one would change on a case-by-case basis, which isn't what
the settings menu is designed for (you get the message, "restart lmms for
changes to take effect").
